### PR TITLE
[SPARK-6051][Streaming] Add ZooKeeper offest posting for DirectKafkaInputDStream

### PR DESCRIPTION
**Pull request by jerryshao**
Currently in DirectKafkaInputDStream, offset is managed by Spark Streaming itself without ZK or Kafka involved, which will make several third-party offset monitoring tools fail to monitor the status of Kafka consumer. So here add an option to post the offset to ZK when each job is finished

### CODE REVIEW

**Comment by tdas** [comment]
@koeninger Would be good to have you in the conversation.

File: project/MimaExcludes.scala [hunk 1]
```diff
@@ -153,6 +153,11 @@ object MimaExcludes {
             ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.RealClock"),
             ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.Clock"),
             ProblemFilters.exclude[MissingClassProblem]("org.apache.spark.TestClock")
+          ) ++ Seq(
+            // SPARK-6051
```
Comment (tdas) [inline project/MimaExcludes.scala]: This should not be required as  as ReliableKafkaReceiver is hidden. But i guess that mima binary checker is perfect.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala [hunk 2]
```diff
@@ -82,8 +83,12 @@ class DirectKafkaInputDStream[
     }
   }
 
+  private val offsetMap = new mutable.HashMap[Time, Map[TopicAndPartition, Long]]()
```
Comment (koeninger) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala]: Isn't this a duplication of information already available in the checkpoint?

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala [hunk 3]
```diff
@@ -158,4 +166,37 @@ class DirectKafkaInputDStream[
     }
   }
 
+  private[streaming]
+  class DirectKafkaStreamingListener extends StreamingListener {
+
+    val offsetCommitEnabled = context.conf.getBoolean("spark.streaming.kafka.offsetCommit", false)
+
+    val zkClientOpt = if (offsetCommitEnabled) {
+      kafkaParams.get("zookeeper.connect").map { zkConnect =>
```
Comment (koeninger) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala]: Doing things this way now means that zookeeper connect is required in the kafka params, whereas it was not previously.

Why not use the existing setConsumerOffset methods in KafkaCluster?  That also means that this could more easily be updated to work with Kafka-based consumer offset storage post-0.8.2

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala [hunk 4]
```diff
@@ -158,4 +166,37 @@ class DirectKafkaInputDStream[
...
+        val zkSessionTimeoutMs = kafkaParams.getOrElse("zookeeper.session.timeout.ms", "6000").toInt
+        val zkConnectionTimeoutMs =
+          kafkaParams.getOrElse("zookeeper.session.timeout.ms", "6000").toInt
+        new ZkClient(zkConnect, zkSessionTimeoutMs, zkConnectionTimeoutMs, ZKStringSerializer)
+      }
+    } else {
+      None
+    }
+
+    val groupId = "directKafkaConsumer"
```
Comment (koeninger) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala]: group id is a kafka param, why hard code it here?

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala [hunk 5]
```diff
@@ -158,4 +166,37 @@ class DirectKafkaInputDStream[
...
+
+    override def onBatchCompleted(batchCompleted: StreamingListenerBatchCompleted) = synchronized {
+      if (offsetCommitEnabled) {
+        for (offsets <- offsetMap.get(batchCompleted.batchInfo.batchTime); client <- zkClientOpt) {
```
Comment (koeninger) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala]: if offsetCommitEnabled, zkClientOpt should always be defined, yes?  Why two separate checks?

From a style point of view, people were complaining about monadic syntax earlier, I don't know that semicolon separated on a single line is an improvement.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/ReliableKafkaReceiver.scala [hunk 6]
```diff
@@ -239,21 +239,7 @@ class ReliableKafkaReceiver[
...
-      } catch {
-        case e: Exception =>
-          logWarning(s"Exception during commit offset $offset for topic" +
-            s"${topicAndPart.topic}, partition ${topicAndPart.partition}", e)
-      }
-
-      logInfo(s"Committed offset $offset for topic ${topicAndPart.topic}, " +
-        s"partition ${topicAndPart.partition}")
-    }
+    SparkKafkaUtils.commitOffset(zkClient, groupId, offsetMap)
```
Comment (koeninger) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/ReliableKafkaReceiver.scala]: Why introduce a separate SparkKafkaUtils object, and how is it different from KafkaUtils?

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/SparkKafkaUtils.scala [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+  def commitOffset(zkClient: ZkClient, groupId: String,
+      offsetMap: Map[TopicAndPartition, Long]): Unit = {
+    if (zkClient == null) {
+      throw new IllegalStateException("Zookeeper client is unexpectedly null")
+    }
+
+    for ((topicAndPart, offset) <- offsetMap) {
+      try {
+        val topicDirs = new ZKGroupTopicDirs(groupId, topicAndPart.topic)
+        val zkPath = s"${topicDirs.consumerOffsetDir}/${topicAndPart.partition}"
```
Comment (koeninger) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/SparkKafkaUtils.scala]: This is relying on details of Kafka's internal usage of ZK paths, as opposed to going through the kafka consumer offset api.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala [hunk 8]
```diff
@@ -158,4 +166,37 @@ class DirectKafkaInputDStream[
     }
   }
 
+  private[streaming]
+  class DirectKafkaStreamingListener extends StreamingListener {
+
+    val offsetCommitEnabled = context.conf.getBoolean("spark.streaming.kafka.offsetCommit", false)
```
Comment (koeninger) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala]: This is a new configuration key, right?

If I'm a new user and see a name like offsetCommit, I'm likely to be confused as to whether the offsets in question are being used as the system of record (they aren't), and whether this is something I have to enable for durability (it isn't).

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala [hunk 9]
```diff
@@ -158,4 +166,37 @@ class DirectKafkaInputDStream[
...
+          kafkaParams.getOrElse("zookeeper.session.timeout.ms", "6000").toInt
+        new ZkClient(zkConnect, zkSessionTimeoutMs, zkConnectionTimeoutMs, ZKStringSerializer)
+      }
+    } else {
+      None
+    }
+
+    val groupId = "directKafkaConsumer"
+
+    override def onBatchCompleted(batchCompleted: StreamingListenerBatchCompleted) = synchronized {
```
Comment (koeninger) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala]: Because ZK is not the system of record, are we really gaining much for the added complexity of onBatchCompleted, as opposed to just updating ZK in compute() ?

**Comment by jerryshao** [comment]
Hi @koeninger , thanks a lot for your review. I will the fix the all the comments you addressed.

The reason why I put updating ZK in `StreamingListener` rather than `compute()` is that `onBatchCompleted` will only be called when batch is completed, whereas `compute()` will be called in each batch duration, so if the job is delayed, lot's of actually not computed but already submitted job will be pending on the queue, but as `compute()` function is already called, so these unprocessed offset will be committed into ZK, so that's why I put ZK updating in `StreamingListener`.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/SparkKafkaUtils.scala (hunk 7 above)
Comment (jerryshao) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/SparkKafkaUtils.scala]: Thanks a lot, I will change to use Kafka consumer offset api.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala (hunk 8 above)
Comment (jerryshao) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala]: So what is your suggestion? shall we set the mechanism as default and make it invisible to user?

Comment (koeninger) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala]: It's more that it needs a clearer name.  Definitely don't think it should be on by default.

Comment (jerryshao) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala]: OK, will change it.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala (hunk 2 above)
Comment (jerryshao) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala]: Do you mean that we can use `batchForTime` ? Looks like this will only be updated when checkpoint is enabled.

**Comment by jerryshao** [comment]
Hi @koeninger , would you please review this again? Thanks a lot and appreciate your time.

Here I still keep using the HashMap for Time -> offset relation mapping, since checkpoint data will only be updated when checkpoint is enabled, I hope this could also be worked even without checkpoint enabled.

And I still use StreamingListener to update the offset, the reason is mentioned before.

Besides I updated the configuration name, not sure is it suitable.

Thanks a lot.

**Comment by koeninger** [comment]
As it stands now, no offsets are stored by spark unless you're
checkpointing.  Does it really make sense to have an option to
automatically store offsets in Kafka, but not store offsets in the
checkpoint?  Failure recovery in that case depends on user provided
starting offsets (or starting at the beginning / end of the log).  If
someone has the sophistication to get offsets from kafka in order to
provide them as a starting point, they probably have the sophistication to
save offsets to kafka themselves in the job.

If offsets are only being sent to Kafka when they are also stored in the
checkpoint, then does sending offsets to kafka in compute() also make
sense?  Yes, you can lag behind, but those offsets are in the queue to get
processed at least once.

I'm not 100% sure on the answer to this, its more a question of desired
behavior, but that's why I brought it up.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala [hunk 10]
```diff
@@ -118,6 +123,7 @@ class DirectKafkaInputDStream[
       context.sparkContext, kafkaParams, currentOffsets, untilOffsets, messageHandler)
 
     currentOffsets = untilOffsets.map(kv => kv._1 -> kv._2.offset)
+    offsetMap += ((validTime, currentOffsets))
```
Comment (koeninger) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala]: Don't all mutations of the offsetMap need to be synchronized?

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala [hunk 11]
```diff
@@ -84,6 +83,11 @@ class DirectKafkaInputDStream[
 
   protected var currentOffsets = fromOffsets
 
+  // Map to manage the time -> topic/partition+offset
+  private val offsetMap = new mutable.HashMap[Time, Map[TopicAndPartition, Long]]()
+  // Add to the listener bus for job completion hook
+  context.addStreamingListener(new DirectKafkaStreamingListener)
+
   @tailrec
   protected final def latestLeaderOffsets(retries: Int): Map[TopicAndPartition, LeaderOffset] = {
```
Comment (koeninger) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala]: Is there a reason to even add the streaming listener if the configuration option isn't turned on?  If the config option isn't on, couldn't you skip the listener and skip adding / removing items to the offset map altogether?

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala (hunk 10 above)
Comment (jerryshao) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala]: I don't think this will be called in another thread, so I think `synchronized` is not required.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala [hunk 12]
```diff
@@ -158,4 +165,23 @@ class DirectKafkaInputDStream[
     }
   }
 
+  private[streaming]
+  class DirectKafkaStreamingListener extends StreamingListener {
+
+    val offsetUpdateEnabled = context.conf.getBoolean("spark.streaming.kafka.offsetUpdate", false)
```
Comment (tdas) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala]: Can this feature be enabled simply by specifying the group.id? Why would any one specify the group.id if they do not want the zookeeper to be updated?

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala (hunk 11 above)
Comment (tdas) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala]: Valid point. And the signal to turn on the updating could simply be the "group.id" setting. Isnt it?

**Comment by tdas** [comment]
This got me thinking. To make this direct stream functionally similar to the older stream, we should also pickup the offset position from the ZK if group.id is specified. In other words, if group.id is specified (as it was needed in the previous stream) in the kafka parameters, we should pickup the offsets from ZK, and then update ZK as we process. 

Does that make sense?

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala (hunk 10 above)
Comment (tdas) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala]: I don't think offsetMap is required at all. The offset ranges are stored in the KafkaRDDs. Why not just pull up the offset from them, rather than introducing another hashmap that needs to be managed/cleared?

**Comment by jerryshao** [comment]
Yes, that's my original purpose of PR, user can monitor and reuse these offsets, also we can offer user the similar functionality as old Kafka stream to get data from previous offset (if not out-of-range and no checkpointing). My concern is that this might break the exact-once semantics. 

So basically let me rephrase what I think about it:

1. If checkpointing is not enabled, do we need to offer user the ability to get data from previous offset or always from earliest or latest offsets? If we want to offer the similar functionality as previous Kafka stream, then I think putting offsets to Kafka is required and keep basic no data loss is necessary.
2. If checkpointing is enabled, these offsets is only used for monitoring, otherwise we will break the exact-once semantics when using these offsets again. So if checkpointing is enabled, we can update the offset at any time because we don't use these offsets to make sure no data loss.

**Comment by tdas** [comment]
If checkpointing is not enabled, then we cannot restart on failure any ways. So the system cannot provide any guarantee. Thats not the point of this update. The point of this PR is to provide the ability to automatically pick up offsets when started (not when restarted) if group.id is specified. Then, if checkpoint is enabled and the context is restarted, it should use the offsets in the checkpoint rather than the offset in ZK. This would give exactly-once semantics under driver failures, as well as the ease of use by just specifying ZK. 

@koeninger What do you think about this?

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala (hunk 10 above)
Comment (koeninger) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala]: I agree that offsetMap has redundant info, but unless checkpointing is turned on, prior offset ranges aren't tracked.

As for the synchronization, if multiple threads aren't touching offsetMap, why do you have the synchronized block in the listener?

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala (hunk 11 above)
Comment (koeninger) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala]: I don't agree that group.id has anything at all to do with tracking offsets in zookeeper.  Its purpose is to identify a group of related consumers.  From the kafka docs:

"A string that uniquely identifies the group of consumer processes to which this consumer belongs. By setting the same group id multiple processes indicate that they are all part of the same consumer group."

Those consumers may or may not be committing offsets to ZK

**Comment by koeninger** [comment]
I am now confused about what the purpose of this PR is.  The jira seemed to indicate that the problem was "several third-party offset monitoring tools fail to monitor the status of Kafka consumer", and that the solution was just to duplicate offsets into ZK, not to use ZK as the system of record for offsets.

Using ZK as the system of record for offsets means that transactional exactly-once semantics is impossible.  I wouldn't be in favor of that change.

I really think the best way to handle this issue is to make it easier for users to interact with Kafka-managed offsets.  For instance, if the kafka cluster code proves to be stable, exposing methods to get and set the consumer offsets.  That way people can choose the semantics they need, by saving offsets to ZK at the point in the job that makes sense for their uses.

That's what we do with our jobs - store offsets into the system of record, then duplicate them into ZK so that existing monitoring tools are happy.

File: external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala (hunk 10 above)
Comment (tdas) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala]: Why even track that redundant info. The offsets can be retrieved as 
```
generatedRDDs.get(batchCompleted.batchInfo.batchTime).asInstanceOf[HasOffsetRanges].offsetRanges
```

Comment (koeninger) [inline external/kafka/src/main/scala/org/apache/spark/streaming/kafka/DirectKafkaInputDStream.scala]: if it's not checkpointed, is rememberDuration guaranteed to be high enough that the rdd hasn't already been cleared out by the time the batch completion listener runs?

If that's all ok, sure, seems much better to me than keeping the redundant info around.

**Comment by tdas** [comment]
@koeninger I agree the purpose of this PR was just to update ZK for tools to work. I am thinking beyond on the bigger goal to make this new API a superset in features of the old API. May be this PR is not the right place to discuss and we can move this discussion to a different location if you feel so. However for the sake of answering your questions let me give the answers here. 

What I was proposing the following. The simpler direct API (the one that does not take explicit offsets) can pick up the initial offset from ZK if group.id is specified. This is exactly similar to how we find the latest or earliest offset in that API, just that instead of latest/earliest offset of the topic we want to find the offset from the consumer group. The group offsets is ZK is not used at all for any further processing and restarting, so the exactly-once semantics is not broken. 

The use case where this is useful is simplified code upgrade. If the user wants to upgrade the code, he/she can the context stop gracefully which will ensure the ZK consumer group offset will be updated with the last offsets processed. Then the new code is started (not restarted from checkpoint) can pickup  the consumer group offset from ZK and continue where the previous code had left off. 

Without the functionality of picking up consumer group offsets to start (that is, currently) the only way to do this is for the users to save the offsets somewhere (file, database, etc.) and manage the offsets themselves. I just want to simplify this process. 

Either way, I am moving this discussion to this JIRA
https://issues.apache.org/jira/browse/SPARK-6249

**Comment by tdas** [comment]
This had fallen off my radar. @jerryshao Lets try to get this PR in for Spark 1.4.0

**Comment by jerryshao** [comment]
Hi @tdas , it's quite a old PR, since @koeninger has several comments on this, I'm not sure which way is the good way :).

**Comment by tdas** [comment]
Can you update with master. Will take a look tomorrow and decide.

**Comment by jerryshao** [comment]
Just rebase the code, I will change the implementation according to feedbacks.

**Comment by koeninger** [comment]
I just want to re-iterate from the jira discussion that I am thumbs-down to the idea of overloading the meaning of group.id, and thumbs-down to adding options that confuse the question of what is the system of record for offsets.

**Comment by zzcclp** [comment]
is there any conclusion?
@jerryshao , could it add a config option to decide whether adding zk offest posting? just my bit suggest:)

**Comment by zzcclp** [comment]
is there any conclusion?

**Comment by jerryshao** [comment]
Hi @zzcclp , would you please describe your requirement and scenario to use this feature. We're still not sure about the necessity of this feature to put offset into ZK/broker.

**Comment by zzcclp** [comment]
Hi @jerryshao , current spark version is not built-in support for upgrade and accumulators when recovering from checkpoint, so I want to add kafka offset to zk, and read offset from zk when restarting or recovering, but not checkpoint.
Recovering from checkpoint has a lot of limitations at current version.

**Comment by koeninger** [comment]
Is there anything you need to do that couldn't be accomplished by reading
from / writing to ZK yourself?

Is this just a question of convenient api for kafka-managed consumer
offsets?

**Comment by zzcclp** [comment]
Hi @koeninger , of cource, I can read kafka offset from ZK when recovering by myself, I just want to know that writing to ZK is built-in support or not.

**Comment by tdas** [comment]
@jerryshao We have now exposed offsets in both Java/Scala/Python API, the user can post the offsets in their own data stores. As @koeninger has pointed out in the JIRA, using the group.id does make the semantics confusing. This is especially important since we want to differentiate as much as possible from the receiver-based approach which uses group.id. So mind closing this PR?

**Comment by jerryshao** [comment]
OK, get it, that makes sense. I'm closing it now.
